### ISSUE TITLE
week #6

### ISSUE THREAD

**Issue by joonsun-baek** [ml-mate/coursera-machine-leaning#11]

**Comment by joonsun-baek** [comment]
### bias, variance

* bias (편향)
  * feature들이 모델에 잘 학습되지 못함을 나타냄
  * bias가 크다는 말은 즉, 모델이 underfitting된 상태라는 것을 의미

* variance (분산)
  * feature들이 모델에 너무 잘 학습된 상태를 나타냄
  * variance가 크다는 말은, overfitting된 상태를 의미

* bias, variance
  * bias된 상태는 training샘플들이 모델에 잘 학습되지 못한 상태로서
    * 학습에 도움을 주는 feature 개수를 늘리거나
    * feature를 polynominal하게 변형해보거나 하는 등의 시도가 가능
  * variance는 overfitting된 상태로서
    * feature 개수를 줄이거나, polynominal상태를 줄이거나,
    * 샘플의 개수를 늘여서 overfitting을 줄이는 시도가 가능

**Comment by joonsun-baek** [comment]
### 모델의 학습 및 평가
* 전형적인 모델의 학습/평가는 주어진 데이터셋을 training(60%), cross validation(20%), test(20%)로 나누어 수행
* cross validation을 활용한 학습 flow
  1. 주어진 데이터셋을 training(60%), cross validation(20%), test(20%)으로 나눔
  2. training 데이터셋으로 `h(x), J(theta), Gradient` 연산
  3. cross validation 데이터셋으로 `Gradient`로부터 `error`를 산출
  4. `error`로 해당 모델의 적합도를 평가

**Comment by joonsun-baek** [comment]
### train/cv셋을 활용한 모델의 진단 1. feature의 polinomial과 cost의 관계를 통한 bias/variance 진단
![스크린샷 2019-08-16 오후 2 01 42](https://user-images.githubusercontent.com/18119192/63145189-6449cf00-c031-11e9-84c4-3eb6053ca325.png)

![스크린샷 2019-08-16 오후 2 01 37](https://user-images.githubusercontent.com/18119192/63145191-6449cf00-c031-11e9-97dd-5e373af020de.png)

---


### train/cv셋을 활용한 모델의 진단 2. lambda와 cost의 관계를 통한 bias/variance 진단
* ![스크린샷 2019-08-16 오후 2 04 44](https://user-images.githubusercontent.com/18119192/63145201-6c097380-c031-11e9-935a-11f45675387a.png)
* ![스크린샷 2019-08-16 오후 2 04 41](https://user-images.githubusercontent.com/18119192/63145202-6c097380-c031-11e9-9a41-c4d54c1f63f2.png)

---

### train/cv셋을 활용한 모델의 진단 3. training set size (m)을 활용한 bias/variance 진단
* high bias
  * training 데이터셋이 적을때는
    * 아직 학습이 잘 안됐으며, cv를 활용해도 학습이 잘 안된상태이기 때문에, cv의 error가 크며, train - cv의 error의 차이가 심함
  * training 데이터셋이 많을때는 
    * 학습이 되어갈수록 train/cv셋 모두의 error가 높게 나옴. 한마디로, bias한 상태
  * ![스크린샷 2019-08-16 오후 2 17 30](https://user-images.githubusercontent.com/18119192/63145213-73308180-c031-11e9-8769-19b0d15e5fe5.png)


* high variance
  * training 데이터셋이 적을때는
    * 아직 학습이 잘 안됐으며, cv를 활용해도 학습이 잘 안된상태이기 때문에, cv의 error가 크며, train - cv의 error의 차이가 심함
  * training 데이터셋이 많을때는 
    * overfit되어 train셋의 error는 낮지만, cv셋의 error는 높음. 한마디로, bias한 상태
  * ![스크린샷 2019-08-16 오후 2 17 32](https://user-images.githubusercontent.com/18119192/63145218-76c40880-c031-11e9-9be9-9075c4ffb974.png)

**Comment by phh0606c** [comment]
CV 에 대해 이해가 잘안되네요... ㅠ

Training Set 으로 모델을 만들고, Test Set 으로 모델의 성능을 평가하는게 지금까지의 방법
이 데이터셋에 적절한 차수(다차함수)를 찾기 위해서, 다차 모델을 만들어서 각각 평가해야함.
이 때, 다차 모델을 예를 들어, 1차함수부터 10차함수 까지 만들어서 모델을 만들었을때, 
몇차 다항식이 가장 적절한가를 확인하기 위하여, CV라는 데이터 셋을 도입. 
(여기서 .. 왜 기존의 Test Set을 쓸수없는지?)
CV 를 통해, 가장 성능이 좋은 모델 다항식을 고르고, 이 모델의 성능 평가는 다시 TEST Set으로 진행. 
(CV도 결국 학습에 사용되지 않는 Test Set 인데, 왜 다시 Test Set으로 성능 평가하는지.)

괄호에 대한 질문에 대한 대답으로 제 생각으로, 성능 평가에 가장 좋은 성능을 뽑은 것을 그 모델의 성능이다. 라고 규정하면 오버피팅일수도 있다고 생각해서 
CV는 적절한 모델의 차수를 찾는데 사용하는 값.
Test Set 은 그 모델의 성능. 
이라고 이해를 했음.

그런데 
![image](https://user-images.githubusercontent.com/22519029/64241364-8783e200-cf3e-11e9-83d1-7cd29b69e6d5.png)
이 문제를 풀고나니. 다시 멘붕 .......

최종적으로 선택된 모델의 Jcv(세타) 는 Jtest(세타) 보다 다음 이유로 값이 작을것으로 기대된다.
..

혼자 정리하는겸 글 적다보니까. 
그냥 헷갈려서 그렇군.. 
Jcv 세타가 값이 작다는게 성능이 더 좋다는 얘기인데, Jcv 성능이 Jtest 성능보다 안좋은것으로 해석해버렸네..  이런 멍청이. ㅋ

**Comment by phh0606c** [comment]
음 Bias Variance 보다가 궁금한점.. 
그래프의 의미는 이해되긴되는데 
결국 적절한 값을 구하는 것은 Jcv가 최소값인 경우를 통해서 구할 수 있다는걸 보여주기 위한 그래프이고, Jtrain 은 어디가 적정선인지 캐치할 수 없다. 는걸 보여주기 위한 그림일까요??
![image](https://user-images.githubusercontent.com/22519029/64246449-c5d1cf00-cf47-11e9-90de-a7b9570b7bb4.png)
![image](https://user-images.githubusercontent.com/22519029/64246466-cbc7b000-cf47-11e9-8df1-6cdcac2383e0.png)

추가로, 순서는 적절한 다항식의 차수를 찾은 다음 적정한 람다(정규화)를 찾는다는 것이겠죠?
해서 최종적으로 생성되는 모델은,
차수와 세타가 정해지도록

**Comment by joonsun-baek** [comment]
움 그랗지
첫번째 그래프에서 Jcv는 underfitting에서 overfitting으로 넘어가는 지점 (Jcv의 최솟값 지점)을 찾아내어,
기존에 Jtraining에선 못찾았던 **적정한 d값**을 찾아낼 수 있다인거고,

두번째 그래프도 **적절한 람다**를 찾는다는.

결국 Jcv는 최적의 **하이퍼 파라미터들**과와 가장 **적절한 m의 사이즈** 등을 찾아내어,
모델이 최적의 세팅값과 데이터셋 사이즈로 활용될 수 있는 방안을 찾아내는데 사용되는 느낌
(예시에선 람다만 활용했는데, 실제로 사용해보니 '가우시안'의 경우 sigma값,
다른 모델들에선 모델에 활용되는 갖가지 하이퍼 파라미터들의 튜닝도 가능한듯!!)

모델 자체의 error는 Jtest로 하는 느낌이고

**Comment by phh0606c** [comment]
![image](https://user-images.githubusercontent.com/22519029/64308622-d9774700-cfd4-11e9-9c2f-c04d8f61c257.png)

모델의 bias 와 variance 를 개선하는 방법

**Comment by phh0606c** [comment]
스팸 분류기를 보면서 한가지 위로가 되는게,
좋은 모델을 위해서 에러 아날리시스 - 수동으로 각각의 이메일을 조사하고, 피처를 보고 스팸인지 아닌지 실제 판단해 보는 리소스를 써라는것. 이것이 도움이 가장 많이 된다.고 하는것. 
근데 위로는 되는데 이거 이렇게 하는데 진짜 시간이 어느정도 걸릴지 감이 안오네..

**Comment by phh0606c** [comment]
와 신기하네. 들은 기억이 전혀없는데 그래프는 본적 있다 싶었는데 들었었구나 ㄷㄷ
헷갈리는 것들이 있었는데 준선이형 다시 이 글을 죽 읽으니 정리가 된다. 👍

**Comment by phh0606c** [comment]
ex5 보다 보니 처음보는 함수가 있다. 
X_poly_test = bsxfun(@minus, X_poly_test, mu);
X_poly_test = bsxfun(@rdivide, X_poly_test, sigma);

https://iamtaehoon.wordpress.com/2015/01/19/matlab-boosting-tip-bsxfun/

음 요소별 계산을 빠르게하기 위한 함수라고 하는데, 
. 연산이랑은 좀 다른건가?
X_poly_test = X_poly_test .- mu
X_poly_test = X_poly_test ./ sigma

![image](https://user-images.githubusercontent.com/22519029/109409614-9e8e7d80-79d7-11eb-80ee-c6fffd51cd76.png)
![image](https://user-images.githubusercontent.com/22519029/109409619-a817e580-79d7-11eb-93a5-d24436274243.png)

확인해보니 동일하다.
성능에 차이가 있나? 아무튼 요소별 연산하는거고, . 으로 연산하는거랑 결과는 같음.
